middleware/timeout.go

```diff
@@ -55,62 +59,163 @@ func TimeoutWithConfig(config TimeoutConfig) echo.MiddlewareFunc {
 	if config.Skipper == nil {
 		config.Skipper = DefaultTimeoutConfig.Skipper
 	}
+	if config.DefaultTimeoutErrorHandler == nil {
+		config.DefaultTimeoutErrorHandler = func(contextError error, c echo.Context) error {
+			return echo.NewHTTPError(http.StatusServiceUnavailable, contextError.Error())
+		}
+	}
+	if config.OnTimeoutRouteErrorHandler == nil {
+		config.OnTimeoutRouteErrorHandler = func(err error, c echo.Context) {
+			if c.Logger() == nil {
+				return
+			}
+			c.Logger().Errorf("handler for path %v returned an error after request timeouted", c.Path())
+		}
+	}
 
 	return func(next echo.HandlerFunc) echo.HandlerFunc {
 		return func(c echo.Context) error {
 			if config.Skipper(c) || config.Timeout == 0 {
 				return next(c)
 			}
 
-			handlerWrapper := echoHandlerFuncWrapper{
-				ctx:        c,
-				handler:    next,
-				errChan:    make(chan error, 1),
-				errHandler: config.OnTimeoutRouteErrorHandler,
+			ctx, cancelCtx := context.WithTimeout(c.Request().Context(), config.Timeout)
+			defer cancelCtx()
+
+			r := c.Request().WithContext(ctx)
+
+			tw := &timeoutWriter{
+				w:   c.Response(),
+				h:   make(map[string][]string),
+				req: r,
 			}
-			handler := http.TimeoutHandler(handlerWrapper, config.Timeout, config.ErrorMessage)
-			handler.ServeHTTP(c.Response().Writer, c.Request())
+			originalWriter := c.Response().Writer
+			c.Response().Writer = tw
+
+			done := make(chan error)
+			defer close(done)
+			panicChan := make(chan interface{})
+			defer close(panicChan)
+
+			go func() {
+				defer func() {
+					if p := recover(); p != nil {
+						panicChan <- p
+					}
+				}()
+
+				err := next(c)
+				// NB: when timeout occurs this functions returns after timeout response has already been sent to the client
+				// only thing we can do with error is to log it as we can not send anything to client anymore
+				select {
+				case <-ctx.Done():
+					config.OnTimeoutRouteErrorHandler(err, c)
+				default:
+					done <- err
+				}
+			}()
 
 			select {
-			case err := <-handlerWrapper.errChan:
+			case p := <-panicChan:
+				panic(p)
```

Comment:
Here we need to set the original writer `c.Response().Writer = originalWriter`. Otherwise we always receive a `200` status code.